### PR TITLE
[android] Adapt POSIX test to Android

### PR DESCRIPTION
**Pull request by drodriguez**
Android differs in subtle ways from Linux. For example `sem_open` is completely unimplemented, and always returns `ENOSYS`. None of the test using `sem_open` were going to work. The test was `XFAIL` for `linux-androideabi`, but that target doesn’t match Android AArch64 (`aarch64-none-linux-android`). Since only the `sem_open` tests will not work, I disabled those individually for Android, instead of disabling the full file.

Additionally the `l_len` field of the flock struct is typed as `__kernel_off_t`, which doesn't have the same size as `off_t`.

Finally, the current working directory is always `/`, so creating relative files will fail. We have to find the executable directory and create the file there. For some reason the `chdir` at the start might work for the parent process, and then doesn't work for the child processes (since `CommandLine.arguments[1]` is actually `--stdlib-unittest-run-child`), but doesn’t fail. The implemented solution should write the files side by side with the executable in Android.

### CODE REVIEW

File: test/stdlib/POSIX.swift [hunk 1]
```diff
@@ -17,21 +13,59 @@ chdir(CommandLine.arguments[1])
...
+    fatalError("Couldn't obtain executable path")
+  }
+
+  let length = strlen(resolvedPath)
+  guard length != 0 else {
+    fatalError("Couldn't obtain valid executable path")
+  }
+
+  // Search backwards for the last /, and turn it into a null byte.
+  for idx in (0..<length).reversed() {
```
Comment (moiseev) [inline test/stdlib/POSIX.swift]: Honestly I can't tell which version simpler to read, but reversing a range looks a bit odd. I'll just mention it here, feel free to ignore :-)

```suggestion
  for idx in stride(from: length-1, through: 0, by: -1) {
```

File: test/stdlib/POSIX.swift [hunk 2]
```diff
@@ -17,21 +13,59 @@ chdir(CommandLine.arguments[1])
...
+  defer {
+    resolvedPath.deinitialize(count: capacity)
+    resolvedPath.deallocate()
+  }
+  guard let _ = realpath("/proc/self/exe", resolvedPath) else {
+    fatalError("Couldn't obtain executable path")
+  }
+
+  let length = strlen(resolvedPath)
+  guard length != 0 else {
```
Comment (moiseev) [inline test/stdlib/POSIX.swift]: I believe this "guard and fatalError" pattern can be expressed by a single call.
```swift
precondition(length != 0, "Couldn't obtain a valid executable path")
```

File: test/stdlib/POSIX.swift [hunk 3]
```diff
@@ -17,21 +13,59 @@ chdir(CommandLine.arguments[1])
...
+  }
+
+  // Search backwards for the last /, and turn it into a null byte.
+  for idx in (0..<length).reversed() {
+      if Unicode.Scalar(UInt8(resolvedPath[idx])) == Unicode.Scalar("/") {
+        resolvedPath[idx] = 0
+        break
+      }
+
+      guard idx != 0 else {
```
Comment (moiseev) [inline test/stdlib/POSIX.swift]: Same comment about using `precondition(_:, _:)` as above.

**Comment by moiseev** [comment]
@swift-ci Please test

**Comment by moiseev** [comment]
@swift-ci Please test

**Comment by drodriguez** [comment]
Failure seems unrelated to the changes.

**Comment by moiseev** [comment]
@swift-ci Please smoke test Linux platform

**Pull request merged**